runners/core-java/src/main/java/org/apache/beam/runners/core/GroupAlsoByWindowViaWindowSetNewDoFn.java

```diff
@@ -0,0 +1,136 @@
+package org.apache.beam.runners.core;
+
+import org.apache.beam.runners.core.triggers.ExecutableTriggerStateMachine;
+import org.apache.beam.runners.core.triggers.TriggerStateMachines;
+import org.apache.beam.sdk.transforms.Aggregator;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.OldDoFn;
+import org.apache.beam.sdk.transforms.Sum;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.util.SideInputReader;
+import org.apache.beam.sdk.util.TimerInternals;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.util.state.StateInternals;
+import org.apache.beam.sdk.util.state.StateInternalsFactory;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+import java.util.Collection;
+
+public class GroupAlsoByWindowViaWindowSetNewDoFn<
+        K, InputT, OutputT, W extends BoundedWindow, RinT extends KeyedWorkItem<K, InputT>>
+        extends DoFn<RinT, KV<K, OutputT>> implements DoFnRunner.ReduceFnExecutor<K, InputT, OutputT, W> {
+
+  private static final long serialVersionUID = 1L;
+
+  public static <K, InputT, OutputT, W extends BoundedWindow>
+      DoFn<KeyedWorkItem<K, InputT>, KV<K, OutputT>> create(
+          WindowingStrategy<?, W> strategy,
+          StateInternalsFactory<K> stateInternalsFactory,
+          TimerInternals timerInternals,
+          SideInputReader sideInputReader,
+          SystemReduceFn<K, InputT, ?, OutputT, W> reduceFn,
+          DoFnRunners.OutputManager outputManager,
+          TupleTag<KV<K, OutputT>> mainTag) {
+    return new GroupAlsoByWindowViaWindowSetNewDoFn<>(
+            strategy, stateInternalsFactory, timerInternals, sideInputReader,
+            reduceFn, outputManager, mainTag);
+  }
+
+  protected final Aggregator<Long, Long> droppedDueToClosedWindow =
+          createAggregator(
+                  GroupAlsoByWindowsDoFn.DROPPED_DUE_TO_CLOSED_WINDOW_COUNTER, Sum.ofLongs());
+  protected final Aggregator<Long, Long> droppedDueToLateness =
+          createAggregator(GroupAlsoByWindowsDoFn.DROPPED_DUE_TO_LATENESS_COUNTER, Sum.ofLongs());
+  private final WindowingStrategy<Object, W> windowingStrategy;
+  private final StateInternalsFactory<K> stateInternalsFactory;
+  private SystemReduceFn<K, InputT, ?, OutputT, W> reduceFn;
+  private TimerInternals timerInternals;
+  private SideInputReader sideInputReader;
+  private DoFnRunners.OutputManager outputManager;
+  private TupleTag<KV<K, OutputT>> mainTag;
+
+  public GroupAlsoByWindowViaWindowSetNewDoFn(
+          WindowingStrategy<?, W> windowingStrategy,
+          StateInternalsFactory<K> stateInternalsFactory,
+          TimerInternals timerInternals,
```

Comment:
I believe you'll need a `TimerInternalsFactory` since timers should also be per key. It would be used in the same manner as the `StateInternalsFactory`.